### PR TITLE
Time system rework and other refactoring

### PR DESCRIPTION
**Pull request by Apceniy**

### CODE REVIEW

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 1]
```diff
@@ -16,81 +28,192 @@ public static String getBeep() {
...
+        try {
+            final String id = from.replaceAll("[^0-9]", "");
+            if (jda != null)
+                user = jda.retrieveUserById(id).complete();
+            if (guild != null)
+                member = guild.retrieveMemberById(id).complete();
+        } catch (final @NotNull IllegalArgumentException e) {
+            channel.sendMessage("Неправильно указан пользователь!").queue();
+        } catch (final @NotNull ErrorResponseException e) {
+            channel.sendMessage("Указан недействительный пользователь!").queue();
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:200]: this breaks build check for literally no reason

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 2]
```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
         return "Б%sп!".formatted(randomElement(letters));
     }
 
-    public static <T> T plural(final long amount, final T first, final T second, final T five) {
+    public static <T> T plural(long amount, final T one, final T two_to_four, final T five_to_zero) {
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:31]:
```suggestion
    public static <T> T plural(long amount, final T one, final T twoToFour, final T fiveToZero) {
```
`Parameter name 'two_to_four' doesn't match regex '[a-z][A-Za-z\d]*'`
`Parameter name 'five_to_zero' doesn't match regex '[a-z][A-Za-z\d]*'`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 3]
```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
...
+        if (amount < 0)
+            amount *= -1;
         if (amount % 10 == 0 || amount % 10 > 4 || amount % 100 / 10 == 1)
-            return five;
+            return five_to_zero;
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:35]:
```suggestion
            return fiveToZero;
```
`Parameter name 'five_to_zero' doesn't match regex '[a-z][A-Za-z\d]*'`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 4]
```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
...
         if (amount % 10 == 1)
-            return first;
-        return second;
+            return one;
+        return two_to_four;
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:38]:
```suggestion
            return one;
        return twoToFour;
```
`Parameter name 'two_to_four' doesn't match regex '[a-z][A-Za-z\d]*'`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 5]
```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
...
+            from = System.currentTimeMillis();
+        final LinkedHashMap<Integer, Integer> durations = new LinkedHashMap<>();
+        final boolean negative = millis < 0;
+        if (negative)
+            millis *= -1;
+
+        // magic
+        int millisOfTheDay = (int) (millis % (1000 * 60 * 60 * 24));
+        millis -= millisOfTheDay;
+        final GregorianCalendar dateFrom = new GregorianCalendar();
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:67]: `Obsolete date-time type 'GregorianCalendar' used`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 6]
```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
...
+        dateFrom.setTimeInMillis(negative ? from - millis : from);
+        final GregorianCalendar dateTo = new GregorianCalendar();
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:69]: `Obsolete date-time type 'GregorianCalendar' used`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 7]
```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
...
+        dateTo.setTimeInMillis(negative ? from : from + millis);
+        millisOfTheDay = Math.round(((float) millisOfTheDay) / 1000);
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:71]:
```suggestion
        millisOfTheDay = Math.round((float) millisOfTheDay / 1000);
```
`Parentheses around '((float) millisOfTheDay)' are unnecessary`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 8]
```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
...
+        for (final int unit : CHECK_ORDER) {
+            int amount;
+            final int max;
+            if (unit >= Calendar.HOUR_OF_DAY) {
+                max = dateFrom.getMaximum(unit) + 1;
+                amount = millisOfTheDay % max;
+                millisOfTheDay /= max;
+            } else {
+                if (dateFrom.get(unit) == dateTo.get(unit))
+                    continue;
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:81]: `'continue' statement`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 9]
```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
...
+                            amount += max - Math.min(max, dateFrom.get(Calendar.DAY_OF_MONTH));
+                        }
+                        durations.put(Calendar.WEEK_OF_YEAR, amount / 7);
+                        amount %= 7;
                         break;
                 }
-                concatLength++;
-                out.insert(0, amount + " " + plural(amount, DURATION_TEXTS[i][vin ? 2 : 1], DURATION_TEXTS[i][3], DURATION_TEXTS[i][4]));
-                seconds -= amount;
             }
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:110]:
```suggestion
                switch (unit) {
                    case Calendar.YEAR -> amount = dateTo.get(unit) - dateFrom.get(unit);
                    case Calendar.MONTH -> {
                        amount = dateTo.get(Calendar.MONTH) - dateFrom.get(Calendar.MONTH);
                        if (amount < 0) {
                            amount += 12;
                            durations.put(Calendar.YEAR, durations.get(Calendar.YEAR) - 1);
                        }
                    }
                    default -> { // DAY_OF_YEAR
                        amount = dateTo.get(Calendar.DAY_OF_MONTH) - dateFrom.get(Calendar.DAY_OF_MONTH);
                        if (amount < 0) {
                            amount = durations.getOrDefault(Calendar.MONTH, 12);
                            final int years = durations.getOrDefault(Calendar.YEAR, 0);
                            if (amount == 12)
                                durations.put(Calendar.YEAR, years - 1);
                            durations.put(Calendar.MONTH, amount - 1);
                            amount = dateTo.get(Calendar.DAY_OF_MONTH);
                            dateTo.add(Calendar.MONTH, -1);
                            max = dateTo.getActualMaximum(Calendar.DAY_OF_MONTH);
                            amount += max - Math.min(max, dateFrom.get(Calendar.DAY_OF_MONTH));
                        }
                        durations.put(Calendar.WEEK_OF_YEAR, amount / 7);
                        amount %= 7;
                    }
                }
```
`Switch statement can be replaced with enhanced 'switch'`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 10]
```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
...
+                durations.put(unit, amount);
+        }
+        
+        // build a string
+        final StringBuilder out = new StringBuilder();
+        int i = 0;
+        String[] names;
+        for (final var entry : durations.entrySet()) {
+            names = DURATION_TYPES.get(entry.getKey()).getRight();
+            out.append((negative && entry.getValue() > 0 ? "-" : "") + entry.getValue() + " " + plural(entry.getValue(), names[accusative ? 1 : 0], names[2], names[3]));
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:128]:
```suggestion
            out.append(negative && entry.getValue() > 0 ? "-" : "").append(entry.getValue()).append(" ").append(plural(entry.getValue(), names[accusative ? 1 : 0], names[2], names[3]));
```
`String concatenation as argument to 'StringBuilder.append()' call`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 11]
```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
...
-
-            buffer = (" " + input).split("\\D+(?=\\d+$)");
+            buffer = (" " + input).split("[^\\d-]+?(?=-?\\d+$)");
             input.setLength(input.length() - buffer[1].length());
-            result += Integer.parseInt(buffer[1]) * multiplier;
+            durations.put(unit, Integer.parseInt(buffer[1]));
         }
-        return result;
+        
+        final Calendar date = new GregorianCalendar();
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:176]: `Obsolete date-time type 'Calendar' used`
`Obsolete date-time type 'GregorianCalendar' used`

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 12]
```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
...
+        date.setTimeInMillis(from);
+        for (final var type : DURATION_TYPES.entrySet()) {
+            if (durations.containsKey(type.getValue().getLeft()))
+                date.add(type.getKey(), durations.remove(type.getValue().getLeft()));
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:180]: `Should be one of: Calendar.ERA, Calendar.YEAR, Calendar.MONTH, Calendar.WEEK_OF_YEAR, Calendar.WEEK_OF_MONTH, ...`

**Review by l1ttleO — CHANGES_REQUESTED** [review]

File: src/main/java/ru/l1ttleO/boyfriend/Actions.java [hunk 13]
```diff
@@ -88,7 +88,11 @@ public static void banMember(final @NotNull MessageChannel channel, final @NotNu
         sendNotification(guild, "%s банит %s на%s за `%s`".formatted(author.getAsMention(), banned.getAsMention(), durationString, reason), silent);
     }
 
-    public static void muteMember(final @NotNull MessageChannel channel, final @NotNull Role role, final @NotNull Member author, final @NotNull Member muted, final String reason, final long duration, final String durationString, final boolean silent) {
+    public static void muteMember(final @NotNull MessageChannel channel, final @NotNull Role role, final @NotNull Member author, final @NotNull Member muted, final String reason, int duration, String durationString, final boolean silent) {
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Actions.java]:
```suggestion
public static void muteMember(final @NotNull MessageChannel channel, final @NotNull Role role, final @NotNull Member author, final @NotNull Member muted, final String reason, final long duration, String durationString, final boolean silent) {
```

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java (hunk 8 above)
Comment (Apceniy) [reply src/main/java/ru/l1ttleO/boyfriend/Utils.java:81]: what's wrong?

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java [hunk 14]
```diff
@@ -16,81 +34,200 @@ public static String getBeep() {
...
+        int millisOfTheDay = (int) (millis % (1000 * 60 * 60 * 24));
+        millis -= millisOfTheDay;
+        final LocalDate dateFrom = LocalDate.ofInstant(Instant.ofEpochMilli(negative ? from - millis : from), ZoneOffset.UTC);
+        LocalDate dateTo = LocalDate.ofInstant(Instant.ofEpochMilli(negative ? from : from + millis), ZoneOffset.UTC);
+        for (final ChronoField unit : DURATION_TYPES.keySet()) {
+            if (unit == ChronoField.ALIGNED_WEEK_OF_YEAR) continue;
+            int amount;
+            final int max;
+            if (unit.ordinal() <= ChronoField.HOUR_OF_DAY.ordinal()) {
+                max = (int) unit.getBaseUnit().getDuration().toMillis();
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/Utils.java:77]: `Cast from 'long' to 'int' may result in loss of precision`

File: src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java [hunk 15]
```diff
@@ -24,8 +24,8 @@ public Remind() {
     public static final @NotNull ThreadGroup REMINDERS_THREAD_GROUP = new ThreadGroup("Reminders");
     public static final @NotNull HashMap<Long, HashMap<DelayedRunnable, Pair<String, Long>>> REMINDERS = new HashMap<>();
 
-    public void run(final @NotNull MessageReceivedEvent event, final @NotNull String @NotNull [] args) throws IntegerOverflowException, InvalidAuthorException, WrongUsageException {
-        final int duration;
+    public void run(final @NotNull MessageReceivedEvent event, final @NotNull String @NotNull [] args) throws NumberOverflowException, InvalidAuthorException, WrongUsageException {
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:27]: `The declared exception 'NumberFormatException' is never thrown`

File: src/main/java/ru/l1ttleO/boyfriend/commands/Mute.java [hunk 16]
```diff
@@ -71,20 +71,20 @@ public void run(final @NotNull MessageReceivedEvent event, final @NotNull String
...
-        String durationString = "всегда";
         if (duration > 0) {
             if (args.length < 4)
                 throw new WrongUsageException("Требуется указать причину!");
-            durationString = " " + Utils.getDurationText(duration, true);
             reasonIndex++;
         } else if (duration < 0)
-            throw new IntegerOverflowException();
+            throw new NumberOverflowException("Введена слишком большая продолжительность, из-за чего она стала отрицательной");
+        else duration = 3600_000;
```
Comment (l1ttleO) [inline src/main/java/ru/l1ttleO/boyfriend/commands/Mute.java:86]: `Group in number literal with underscores does not have length 3`

**Review by l1ttleO — CHANGES_REQUESTED** [review]

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java (hunk 14 above)
Comment (Apceniy) [reply src/main/java/ru/l1ttleO/boyfriend/Utils.java:77]: This line is called for hours, minutes and seconds. For hours it's 3600000, for minutes it's 60000, for seconds it's 1000. Precision can't be lost there.

File: src/main/java/ru/l1ttleO/boyfriend/commands/Mute.java (hunk 16 above)
Comment (Apceniy) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Mute.java:86]: I used it to separate seconds and milliseconds

File: src/main/java/ru/l1ttleO/boyfriend/Utils.java (hunk 14 above)
Comment (l1ttleO) [reply src/main/java/ru/l1ttleO/boyfriend/Utils.java:77]: https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#toIntExact-long-

File: src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java (hunk 15 above)
Comment (Apceniy) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:27]: Did you mean 'NumberOverflowException'?

Comment (l1ttleO) [reply src/main/java/ru/l1ttleO/boyfriend/commands/Remind.java:27]: ...yes

**Review by l1ttleO — APPROVED** [review]
LGTM

**Pull request merged**